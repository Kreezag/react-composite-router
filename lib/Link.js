/* eslint-env browser */

import React, { PropTypes, Component } from 'react';
import connectRouter from './connect';
import referringRouter, { statePropTypes } from './referring';
import { router as routerPropTypes } from './propTypes';
import { compose } from './utils';




const isLeftClickEvent = (event) =>
  event.button === 0;



const isModifiedEvent = (event) =>
  Boolean(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);



class Link extends Component {
  static propTypes = {
    state: statePropTypes,
    target: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    className: PropTypes.string,
    activeClass: PropTypes.string,
    disabledClass: PropTypes.string,
    activeStateClass: PropTypes.string
  };

  static defaultProps = {
  };

  static contextTypes = {
    router: routerPropTypes.isRequired
  };

  constructor (...args) {
    super(...args);

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick (event) {
    const {
      onClick,
      target,
      disabled,
      state
    } = this.props;

    if (onClick) {
      onClick(event);
    }

    if (
      !event.defaultPrevented
      && !target
      && !isModifiedEvent(event)
      && isLeftClickEvent(event)
    ) {
      event.preventDefault();

      if (!disabled) {
        state.apply();
      }
    }
  }

  render () {
    const {
      state,
      activeClass,
      activeStateClass,
      disabledClass,
      ...otherProps
    } = this.props;

    const className = [
      this.props.className,
      state.isActive && activeClass,
      state.isActiveState && activeStateClass,
      this.props.disabled && disabledClass
    ].filter(Boolean).join(' ');

    return (
      <a
        {...otherProps}
        href={state.href}
        className={className}
        onClick={this.handleClick}
      />
    );
  }
}


export default compose(
  connectRouter(),
  referringRouter()
)(Link);
