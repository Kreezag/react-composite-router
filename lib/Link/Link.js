import { PropTypes } from 'react';
import { compose, getContext } from 'recompose';
import { connect } from 'react-redux';
import LinkComponent from './Link.jsx';
import historyPropTypes from '../util/historyPropTypes';




export default compose(
  getContext({
    history: historyPropTypes(),
    routerRoutesByName: PropTypes.object.isRequired
  }),
  connect(({ routerState }, { history, routerRoutesByName }) => ({
    history,
    routerState,
    routerRoutesByName
  }), {})
)(LinkComponent);
