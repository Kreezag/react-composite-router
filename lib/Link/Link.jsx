/* eslint-env browser */

import React, { PropTypes } from 'react';
import { calcLocation, moveTo } from '../routing';
import historyPropTypes from '../historyPropTypes';


const isLeftClickEvent = (event) =>
event.button === 0;



const isModifiedEvent = (event) =>
  Boolean(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);



const Link = (props) => {
  const {
    routerState,
    routerRoutesByName,
    history,
    state,
    className,
    disabled,
    children,
    reload,
    onClick,
    replace,
    target,
    activeClass,
    activeStateClass,
    disabledClass,
    params = {},
    ...otherProps
  } = props;

  const location = calcLocation(routerRoutesByName, state, routerState.params, params);
  const isActive = history.location.pathname === location.pathname;
  const isActiveState = routerState.name === state || routerState.name.startsWith(state);
  const href = history.createHref(location);

  const handleClick = (event) => {
    if (onClick) {
      onClick(event);
    }

    if (
      !reload
      && !event.defaultPrevented
      && !target
      && !isModifiedEvent(event)
      && isLeftClickEvent(event)
    ) {
      event.preventDefault();

      moveTo(history, location, { replace, reload });
    }
  };

  const classNames = [
    className,
    isActive && activeClass,
    isActiveState && activeStateClass,
    disabled && disabledClass
  ].filter(Boolean).join(' ');

  return (
    <a
      {...otherProps}
      href={href}
      target={target}
      className={classNames}
      disabled={disabled}
      onClick={handleClick}
    >
      {typeof children === 'function' ? children({ isActive, disabled, isActiveState }, routerState) : children}
    </a>
  );
};



Link.propTypes = {
  state: PropTypes.string.isRequired,
  params: PropTypes.object,
  reload: PropTypes.bool.isRequired,
  target: PropTypes.oneOf([ '_blank' ]),
  replace: PropTypes.bool.isRequired,
  onClick: PropTypes.func,
  history: historyPropTypes(),
  disabled: PropTypes.bool.isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  activeClass: PropTypes.string,
  routerState: PropTypes.object.isRequired,
  disabledClass: PropTypes.string,
  activeStateClass: PropTypes.string,
  routerRoutesByName: PropTypes.object.isRequired
};



Link.defaultProps = {
  reload: false,
  replace: false,
  disabled: false
};



Link.contextTypes = {
};



export default Link;
