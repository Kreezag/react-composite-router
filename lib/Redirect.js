/* eslint-env browser */

import React, { Component } from 'react';
import referringRouter, { statePropTypes } from './referring';



const ssr = typeof window !== 'object';



class Redirect extends Component {

  static propTypes = {
    state: statePropTypes
  };

  render () {
    return <noscript />;
  }

  componentWillMount () {
    if (ssr) {
      this.props.state.apply();
    }
  }

  componentDidMount () {
    if (!ssr) {
      this.props.state.apply();
    }
  }
}



export default referringRouter()(Redirect);
