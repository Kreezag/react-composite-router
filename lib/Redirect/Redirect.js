/* eslint-env browser */

import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { moveTo, calcLocation } from '../routing';
import historyPropTypes from '../historyPropTypes';



const ssr = typeof window !== 'object';



class Redirect extends Component {
  static contextTypes = {
    routerRoutesByName: PropTypes.object.isRequired,
    history: historyPropTypes()
  };

  static propTypes = {
    routeState: PropTypes.object.isRequired,
    state: PropTypes.string.isRequired,
    params: PropTypes.object,
    replace: PropTypes.bool,
    reload: PropTypes.bool
  };

  static defaultProps = {
    replace: true,
    reload: false
  };

  render () {
    return null;
  }

  componentWillMount () {
    if (ssr) {
      this._redirect();
    }
  }

  componentDidMount () {
    if (!ssr) {
      this._redirect();
    }
  }

  _redirect () {
    const { history, routerRoutesByName, routerState } = this.context;
    const { replace, reload, state, params } = this.props;

    const location = calcLocation(routerRoutesByName, state, routerState.params, params);

    moveTo(history, location, { replace, reload });
  }
}



export default connect(({ routerState }) => ({ routerState }))(Redirect);
