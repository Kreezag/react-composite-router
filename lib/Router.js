import React, { Component } from 'react';
import isFunction from 'lodash/isFunction';
import PropTypes from 'prop-types';
import RouterContext from './RouterContext';
import { router as routerPropTypes, routes as routesPropTypes } from './propTypes';



class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func,
    routes: PropTypes.oneOfType([ routesPropTypes.isRequired, PropTypes.shape({ getRoutes: PropTypes.func.isRequired }) ]).isRequired,
    history: PropTypes.shape({ // instance of history.js
      action: PropTypes.string,
      location: PropTypes.shape({
        pathname: PropTypes.string,
        search: PropTypes.string,
        hash: PropTypes.string
      }).isRequired,
      push: PropTypes.func.isRequired,
      listen: PropTypes.func.isRequired,
      replace: PropTypes.func.isRequired,
      createHref: PropTypes.func.isRequired
    })
  };

  static childContextTypes = {
    router: routerPropTypes
  };

  getChildContext () {
    return { router: this._router };
  }

  constructor (...args) {
    super(...args);

    this._router = new RouterContext();
  }

  UNSAFE_componentWillMount () {
    const { history, routes, onChange } = this.props;

    this._router.onStateChange((name, params) => {
      if (onChange) {
        onChange(name, params);
      }

      this.forceUpdate();
    });

    if (routes && isFunction(routes.getRoutes)) {
      this._router.setRoutes(routes.getRoutes());
    } else {
      this._router.setRoutes(routes);
    }

    if (history) {
      this._router.setHistory(history);
    }
  }

  componentWillUnmount () {
    this._router.destroy();
    this._router = null;
  }

  render () {
    return this.props.children;
  }
}



export default Router;
