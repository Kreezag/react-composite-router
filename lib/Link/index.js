/* eslint-env browser */

import Link from './Link';



export default Link;
