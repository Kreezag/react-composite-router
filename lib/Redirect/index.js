/* eslint-env browser */

import React from 'react';
import Redirect from './Redirect';



export const redirectRender = (state, params = {}) => () => (<Redirect state={state} params={params} />); // eslint-disable-line



export default Redirect;
