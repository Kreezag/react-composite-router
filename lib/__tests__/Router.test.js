/* eslint-env browser, jest */
/* eslint-disable */

import React from 'react';
import Router from '../Router';
import route from '../route';
import Slot from '../Slot';
import Redirect from '../Redirect';
import Link from '../Link';
import renderer from 'react-test-renderer';
import WithContext from 'react-with-context';
import store from './__mocks__/redux-store';
import history from './__mocks__/history';

describe('rendering', () => {
  it('should render something', () => {
    route.create('state', { url: 'state' });

    const component = renderer.create(
      <WithContext context={{
        store: store({ routerState: { name: '', initialized: true, params: {} } })
      }}>
        <Router history={history()} name="some">
          <div>
            <span>Hello!</span>
            <Link state="state">Some</Link>
            <Redirect state="state"/>
            <Slot name="some" />
          </div>
        </Router>
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
