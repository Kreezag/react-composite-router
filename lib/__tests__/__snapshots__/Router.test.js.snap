exports[`rendering should render something 1`] = `
<div>
  <span>
    Hello!
  </span>
  <a
    className=""
    href="state"
    onClick={[Function]}>
    Some
  </a>
  <noscript />
</div>
`;
